### PR TITLE
perf: GPU-side attention for LLaMA decode (+ CPU loop bounds-check drop)

### PR DESCRIPTION
**Pull request by bararchy**
Moves the LLaMA KV-cache attention head loop onto the GPU. It was ~47% of decode time, pure CPU, and grew with context — the reason inference showed 100% CPU / 25-30% GPU. Also includes the earlier bounds-check removal commit for the (now fallback) CPU loop.

## Commit 1: CPU loop bounds-check drop
Raw-pointer hot loops in `attention_cached_cpu` (~20% faster long decode: 1488→942ms attn time over 192 tok). This loop is now the CPU fallback path.

## Commit 2: GPU-side KV-cache attention
- **Device-resident KV cache** per block, laid out `[num_kv_heads, capacity, head_dim]`, capacity doubling; on growth the CPU mirror (still source of truth) is re-uploaded.
- **Per layer/token**: one staging H2D upload (new RoPE'd K rows + V rows + RoPE'd Q), then two kernels:
  - `kv_cache_append_f32` — scatters new K/V into the device cache;
  - `attention_kv_f32` — fused scores → stable softmax → weighted-V per (head, token) block, GQA-aware (`kv_h = h / heads_per_kv`), causal (`visible = start_pos + i + 1`, holds for prefill M>1 too).
- Only the attention output `[new_tokens, d_model]` comes back to host; w_o projection unchanged.
- **Fallback**: CPU path auto-selected when CUDA kernels unavailable, or forced via `force_cpu_attention` property / `SHAINET_CPU_ATTENTION=1`.
- Persistent grow-only device buffers — no per-token cudaMalloc/free in the hot path.

## Numbers (Llama 3.2 1B Instruct, Q8, greedy, RTX 4090 Laptop)
| | CPU attention | GPU attention |
|---|---|---|
| 192-tok decode | 76.7 tok/s | **102.5 tok/s** |
| 64-tok decode | — | 104.8 tok/s (flat vs context) |
| GPU util during decode | 25-30% | **~65%** |

Output text unchanged: greedy "What's the largest dog breed?" → Irish Wolfhound text, byte-identical at 64 and 256 tokens.

## Testing
- New `spec/llama_attention_gpu_spec.cr`: GPU vs CPU attention on identical weights/inputs — prefill (M>1), incremental decode, GQA mapping, device cache growth past initial capacity, and `clear_cache!` reuse. Max-abs-diff < 1e-3, cosine > 0.9999.
- `crystal spec`: 149 examples, 0 failures.
- CUDA specs (`quantized_cuda_matrix`, `llama_loader`, `cuda_matrix`, `llama_attention_gpu` with `-Denable_cuda`): 9 examples, 0 failures.
- Both builds compile: with and without `-Denable_cuda`.

Note: rebuild kernels after checkout — `nvcc -shared --compiler-options=-fPIC -O3 -arch=sm_89 -I/opt/cuda/include -L/opt/cuda/lib64 -lcurand src/shainet/native/cuda_kernels.cu -o libshainet_cuda_kernels.so`

## Commit 3: llama_chat sampler — O(V) top-k selection
The chat example built a 128K-tuple array and full-sorted the whole vocab every token to pick top-40 (~16ms/tok of pure CPU — chat ran 38 tok/s while greedy q8_eval hit ~100). Replaced with a single-pass partial selection into a sorted 40-element array; identical sampling semantics. 38.0 → 97.9 tok/s.

### CODE REVIEW

**Comment by bararchy** [comment]
Folded in a reusable **`SHAInet::Sampler`** (commit 27b13c4) and rewired both examples to it:

- `Sampler.greedy(logits, row)` — NaN/Inf-safe argmax (stateless).
- `Sampler#sample(logits, row)` — temperature + top-k via O(vocab) partial selection (no 128K sort), stable softmax, inverse-CDF; `top_k=1` == greedy; takes an `rng` for deterministic tests.
- `Sampler#apply_repetition_penalty!(logits, recent_ids, window:)` — in-place `v>0 ? v/p : v*p`.

Fully optimized for decode: instance reuses its top-k buffers (zero per-token allocation in steady state) and hot loops use raw `SimpleMatrix#data` pointers (no per-element bounds checks). `llama_chat` delegates to it (no more inline sort); `q8_eval` uses `Sampler.greedy` with **byte-identical** output.

Verified: 159 specs pass (incl. new `spec/sampler_spec.cr`); `q8_eval` Irish Wolfhound text unchanged @ ~100 tok/s; `llama_chat` coherent @ ~100 tok/s; both builds compile with/without `-Denable_cuda`.
